src/rules/scss-function-pattern/index.js

```diff
@@ -0,0 +1,36 @@
+import { isRegExp, isString } from "lodash"
+import { utils } from "stylelint"
+
+export const ruleName = "scss-function-pattern"
+
+export const messages = utils.ruleMessages(ruleName, {
+  expected: "Expected SCSS function name to match specified pattern",
+})
+
+export default function (pattern) {
+  return (root, result) => {
+    const validOptions = utils.validateOptions(result, ruleName, {
+      actual: pattern,
+      possible: [ isRegExp, isString ],
+    })
+    if (!validOptions) { return }
+
+    const regexpPattern = (isString(pattern))
+      ? new RegExp(pattern)
+      : pattern
+
+    root.walkAtRules(declaration => {
+      const decl = declaration
```

Comment:
I think that the code would be a bit more readable if you could just change `declaration` to `decl` and do something like this:

```js
root.walkAtRules(decl => {
  if (decl.name !== "function") { return }

  const funcName = decl.params.replace(/(\s*?)\(.*?\)/g, "")
  if (regexpPattern.test(funcName)) { return }
```